rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/TextDocumentState.java

```diff
@@ -30,65 +30,78 @@
 
 import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
 import org.rascalmpl.values.parsetrees.ITree;
+import org.rascalmpl.vscode.lsp.util.Versioned;
 
 import io.usethesource.vallang.ISourceLocation;
 
 /**
- * TextDocumentState encapsulates the current contents of every open file editor, 
+ * TextDocumentState encapsulates the current contents of every open file editor,
  * and the corresponding latest parse tree that belongs to it.
  * It is parametrized by the parser that must be used to map the string
- * contents to a tree. All other TextDocumentServices depend on this information. 
- * 
+ * contents to a tree. All other TextDocumentServices depend on this information.
+ *
  * Objects of this class are used by the implementations of RascalTextDocumentService
- * and ParametricTextDocumentService. 
+ * and ParametricTextDocumentService.
  */
 public class TextDocumentState {
     private final BiFunction<ISourceLocation, String, CompletableFuture<ITree>> parser;
 
     private final ISourceLocation file;
-    private volatile String currentContent;
+    private volatile Versioned<String> currentContent;
     @SuppressWarnings("java:S3077") // we are use volatile correctly
-    private volatile @MonotonicNonNull ITree lastFullTree;
+    private volatile @MonotonicNonNull Versioned<ITree> lastFullTree;
     @SuppressWarnings("java:S3077") // we are use volatile correctly
-    private volatile CompletableFuture<ITree> currentTree;
+    private volatile CompletableFuture<Versioned<@Nullable ITree>> currentTree;
 
-    public TextDocumentState(BiFunction<ISourceLocation, String, CompletableFuture<ITree>> parser, ISourceLocation file, String content) {
+    public TextDocumentState(BiFunction<ISourceLocation, String, CompletableFuture<ITree>> parser, ISourceLocation file, int version, String content) {
         this.parser = parser;
         this.file = file;
-        this.currentContent = content;
-        currentTree = newContents(content);
+        this.currentContent = new Versioned<>(version, content);
+        this.currentTree = newContent(version, content);
     }
 
-    public CompletableFuture<ITree> update(String text) {
-        currentContent = text;
-        currentTree = newContents(text);
+    /**
+     * The current call of this method guarantees that, until the next call,
+     * each intermediate call of `getCurrentTreeAsync` returns (a future for) a
+     * *correct* versioned tree. This means that:
+     *   - the version of the tree is parameter `version`;
+     *   - the tree is produced by parsing parameter `content`.
+     *
+     * Thus, callers of `getCurrentTreeAsync` are guaranteed to obtain a
+     * consistent <version, tree> pair.
+     */
+    public CompletableFuture<Versioned<ITree>> update(int version, String content) {
+        currentContent = new Versioned<>(version, content);
+        currentTree = newContent(version, content);
```

Comment:
if we store the result of the `newContent` in a local variable, and assign that to `currentTree` and then return it, it's more clear that you will always get the right `tree` back. (instead of maybe `currentTree` having been updated again).